README.md
# Mar.stranal
### (an R-based version of STRANAL)
#### Dec 15, 2017
---
## Introduction and Approach
STRANAL (or "Stratified Analysis") has been used by Fisheries Biologists at DFO and NMFS to assist with assessments since before 2006.  While the original <a href="https://en.wikipedia.org/wiki/APL_(programming_language)">APL</a> application still works well, the number of people who are able to modify and enhance it is dwindling.  

As a result, we have converted the application to an R package, a format and language known to many within the Population Ecology Division.   By producing a package rather than a script, we are better able to version control the code and ensure consistent results. It should run regardless of OS, working directory, etc...

The hope is that users will be better able to collaborate and enhance Mar.stranal now that the code is readily accessible.  Additionally, since it is now a package rather than an application, its operation is transparent, and PED staff can maintain it without IMTS support.

## <a name="Limitations"></a>  Current Limitations
As of right now, the STRANAL does not yet include:

*  Vessel Net Conversion
*  ALK Modifications (including User-specified age/length table)

## <a name="Running"></a>  Installing and Running
Mar.stranal has a single function - `stranal()` after you install and load the package.

It can be run via command line or via the built-in Graphical User Interface (GUI).  If a correct Oracle username and password are not provided, you will be prompted to provide them.  

If it is run by command line - each input parameter is compared with valid values. If they are not determined to be valid, a select box will present the user with valid options.

You can install Mar.stranal from github with:

```R
# install.packages("devtools")

devtools::install_github("Maritimes/Mar.utils")
devtools::install_github("Maritimes/Mar.stranal")
```

The script can be run like the following:

```R
# very basic - lots of prompts
results <- stranal()

# all details provide - no prompts
results <- stranal(year = 1990, season = "SUMMER", bySex=TRUE, ageBySex = TRUE, missions = c('NED1990140'), spp=14)
```
Depending on what parameters are provided, when the script is run, the user will be prompted for the following:
*  usepkg - "rodbc" or "roracle" - however you connect to oracle 
*  agency - "DFO" or "NMFS" 
*  type = 1
*  year = NULL
*  season = NULL
*  missions = NULL
*  strataTable = "GROUNDFISH.GSSTRATUM""
*  wingspread = 41
*  towDist = 1.75
*  strata = c(440:495)
*  spp = NULL
*  bySex = NULL
*  ageBySex = FALSE
*  output = "new" or "classic". Set as "" to skip excel file generation
  

NEWS
Mar.stranal 2019.04.30
==============

Changes:
2019.05.03
* fixes issue with strata selections not honouring area choice
2019.04.30
* fixes NMFS strata identification
* adds version check
2018.10.17
* Fixes issues with NMFS strata selection
2017.12.15 
* Mar.stranal is now a thing


R/extractData.R
# @title extractData
# @description This function doesn't do anything yet.
# @param requested  The default value is \code{NULL}. This determines which 
# analytic will be performed.
# @param agency  The default value is \code{NULL}. 
# @param dfSpp  The default value is \code{NULL}.
# @param type  The default value is \code{NULL}.
# @param missions  The default value is \code{NULL}.
# @param strata  The default value is \code{NULL}.
# @param bySex  The default value is \code{NULL}.
# @family Gale-force
# @author  Mike McMahon, \email{Mike.McMahon@@dfo-mpo.gc.ca}
#' @keywords internal
extractData<-function(requested = NULL, agency = NULL, dfSpp = NULL, type=NULL,
                      missions = NULL, strata = NULL, areas= NULL, bySex = NULL, 
                      oracle_cxn = NULL){
  areaTweak = ""
  
  getCatch<-function(agency, dfSpp, missions, strata, areas){
    spp=dfSpp$SPEC
    if (agency =="DFO"){
      if (!("all" %in% areas))areaTweak = paste0("I.AREA IN (",Mar.utils::SQL_in(areas),") AND")
      sql <-
        paste("select C.mission,C.setno,C.size_class,C.totwgt,C.sampwgt,C.totno,C.calwt
                from
                groundfish.gscat c,
                groundfish.gsinf i
                where
                C.MISSION = I.Mission AND
                C.SETNO = I.SETNO AND
                C.SPEC =", spp," AND
                ",areaTweak,"
                C.mission IN (",Mar.utils::SQL_in(missions),") AND 
                I.STRAT IN (",Mar.utils::SQL_in(strata),") 
                ORDER BY C.mission,C.setno
                ", sep="")
    }else if (agency=="NMFS"){
      if (!("all" %in% areas))areaTweak = paste0("S.AREA IN (",Mar.utils::SQL_in(areas),") AND")
      #adding fake values for calwt(0), size_class(1), and sampwgt(0) so data format matches CDN data
      sql <-
        paste("select I.cruise6 mission,to_number(I.station) setno, 1 size_class, sum(I.expcatchwt) totwgt, 0 sampwgt, sum(I.expcatchnum) totno, 0 calwt
                from
                usnefsc.uss_catch I,
                usnefsc.uss_station S
                WHERE
                I.ID = S.ID AND 
                ",areaTweak,"
                to_number(I.svspp)=", spp," AND
                I.cruise6 IN (",Mar.utils::SQL_in(missions),") AND
                I.STRATUM   IN (",Mar.utils::SQL_in(strata),")
                group by
                I.cruise6, to_number(I.station)
                ORDER BY I.cruise6, to_number(I.station)", sep="")
    }
    raw.gscat<-oracle_cxn$thecmd(oracle_cxn$channel, sql)
    if (nrow(raw.gscat)<1) stop("Error: No catch data can be found for your selection")
    return(raw.gscat)
  }
  getInf<-function(agency, missions, strata, type, areas){
    
    if (!("all" %in% areas))areaTweak = paste0("I.AREA IN (",Mar.utils::SQL_in(areas),") AND")
    if (agency=='DFO'){
      sql<-
        paste("select i.mission, i.setno,i.sdate,i.time,i.strat,
          i.area unit_area,
i.slat,
i.slong,
i.dmin,i.dmax,i.depth, i.dur,i.dist
          from 
          groundfish.gsinf i 
          where 
          ",areaTweak,"
          i.MISSION IN (",Mar.utils::SQL_in(missions),") AND
          i.strat IN (",Mar.utils::SQL_in(strata),") AND 
          i.type IN (",type,")
          ORDER BY i.mission, i.setno", sep="")
    }else if (agency=="NMFS"){
      #distance was assumed to be 1.75, but appears to be dopdistb
      sql<-
        paste("SELECT CRUISE6 mission,to_number(station) setno, begin_est_towdate sdate, est_time time, STRATUM strat, 
          area unit_area,  BEGLAT slat, BEGLON slong, mindepth dmin, maxdepth dmax, avgdepth depth, towdur dur, dopdistb dist 
          FROM USNEFSC.USS_STATION I
          WHERE 
          ",areaTweak,"
          CRUISE6 IN (",Mar.utils::SQL_in(missions),") AND
          STRATUM IN (",Mar.utils::SQL_in(strata),") AND
          to_number(SHG) <= ",type," 
          ORDER BY CRUISE6,to_number(station)", sep="")
    }

    raw.gsinf<-oracle_cxn$thecmd(oracle_cxn$channel, sql )
    #if (agency=="NMFS") raw.gsinf$STRAT<-sprintf("%05d", raw.gsinf$STRAT)
   
      raw.gsinf$SLAT = (as.numeric(substr(raw.gsinf$SLAT,1,2))+(raw.gsinf$SLAT - as.numeric(substr(raw.gsinf$SLAT,1,2))*100)/60)
      raw.gsinf$SLONG = (as.numeric(substr(raw.gsinf$SLONG,1,2))+(raw.gsinf$SLONG - as.numeric(substr(raw.gsinf$SLONG,1,2))*100)/60)*-1
   
    return(raw.gsinf)
  }
  getDet<-function(agency, missions, strata, dfSpp, bySex, type, areas){
    spp=dfSpp$SPEC
    sppLgrp = dfSpp$LGRP
    #orig stratisfy is inconsistent with berried females:
    # -if analysis is done by sex, berried females (fsex = 3) are ignored
    # -if analysis is done not by sex, they are included
    # The following block does this so that the results match, but it seems 
    # like a bad idea
    if(bySex){
      sexFilt = " (d.FSEX IS NULL OR d.fsex IN (0,1,2)) AND "
    }else{
      sexFilt = ""
    }
    if (agency=='DFO'){
      # APL did NOT limit det results by strata, or type
      sql<-paste("select d.mission,d.setno,d.size_class,
      nvl(d.fsex,0) fsex,d.age, d.fwt,         
       decode(",sppLgrp,",1,d.flen,2,.5+2*floor(d.flen/2),3,1+3*floor(d.flen/3),flen) flen,          
       s.lgrp binwidth,s.lfsexed bysex, d.clen          
       from          
       groundfish.gsdet d, groundfish.gsspec s, groundfish.gsinf i 
       WHERE 
       s.spec=d.spec AND  
       d.MISSION = i.MISSION AND
       d.SETNO = i.SETNO AND 
      -- (d.FSHNO IS NOT NULL OR d.FWT IS NOT NULL) AND
       d.MISSION IN (",Mar.utils::SQL_in(missions),") AND 
       -- i.strat IN (",Mar.utils::SQL_in(strata),") AND
       -- i.area IN (",Mar.utils::SQL_in(areas),") AND
       -- i.type IN (",type,") AND
       ",sexFilt,"
       d.SPEC=",spp," 
       ORDER BY d.mission,d.setno", sep="")
      raw.gsdet<-oracle_cxn$thecmd(oracle_cxn$channel, sql )
      
      #since berried and normal females get combined, add the CLEN for sets
      # raw.gsdet = 
      #   stats::aggregate(list(CLEN = raw.gsdet$CLEN),
      #             by = list(MISSION=raw.gsdet$MISSION,
      #                       SETNO =raw.gsdet$SETNO,
      #                       SIZE_CLASS =raw.gsdet$SIZE_CLASS,
      #                       FSEX =raw.gsdet$FSEX,
      #                       #AGE =raw.gsdet$AGE,
      #                       #FWT =raw.gsdet$FWT,
      #                       FLEN =raw.gsdet$FLEN,
      #                       BINWIDTH =raw.gsdet$BINWIDTH,
      #                       BYSEX =raw.gsdet$BYSEX
      #                       ),
      #             FUN = sum, na.rm=TRUE)
      
    }else if (agency=="NMFS"){
      #missing fsex, sizeclass,clen
      sql1<- paste("select cruise6 mission, to_number(station) setno, length, 
        age, avg(indwt) fwt,
        decode(", sppLgrp,",1,length,2,.5+2*floor(length/2),3,1+3*floor(length/3),length) flen,
        ",sppLgrp," binwidth
        from usnefsc.uss_detail I
        where 
        to_number(svspp)=",spp," AND
        CRUISE6 IN (",Mar.utils::SQL_in(missions),") AND 
        STRATUM IN (",Mar.utils::SQL_in(strata),") 
        group by cruise6,station,age,length
        ORDER BY cruise6, to_number(station)", sep="")
      raw.gsdet1<-oracle_cxn$thecmd(oracle_cxn$channel, sql1 )
      
      sql2<- paste("select cruise6 mission, catchsex fsex, to_number(station) setno,length, 
        sum(expnumlen) clen, 1 size_class
        from usnefsc.uss_lengths
        where to_number(svspp)=",spp,"
        and cruise6 in (",Mar.utils::SQL_in(missions),")
        and catchsex in ('0','1','2')
        group by cruise6,station,length, catchsex
        ORDER BY cruise6, to_number(station)",sep="")
      
      raw.gsdet2<-oracle_cxn$thecmd(oracle_cxn$channel, sql2 )
      raw.gsdet<-merge(raw.gsdet1,raw.gsdet2, all.x=T) 
      raw.gsdet$FLEN[is.na(raw.gsdet$FLEN)] <- raw.gsdet$LENGTH[is.na(raw.gsdet$FLEN)]
    }
    return(raw.gsdet)
  }
  
  switch(requested, 
         "catch" = getCatch(agency,dfSpp,missions, strata, areas), 
         "inf" = getInf(agency, missions, strata, type, areas),
         "det" = getDet(agency, missions, strata, dfSpp, bySex, type, areas)
  )
}

R/getAlkTable.R
# @title getAlkTable
# @description This function doesn't do anything yet.
# @param requested  The default value is \code{NULL}. This determines which 
# analytic will be performed.
# @param alkTable  The default value is \code{NULL}. 
# @param oracle_cxn  The default value is \code{NULL}. 
# @importFrom utils read.csv
# @importFrom utils read.delim
# @importFrom openxlsx read.xlsx
# @family Gale-force
# @author  Mike McMahon, \email{Mike.McMahon@@dfo-mpo.gc.ca}
getAlkTable<-function(alkTable=NULL, oracle_cxn = NULL){
  custAlkTable<-NA
  
  fn<-basename(alkTable)
  fn_dets = strsplit(fn, '[.]')[[1]]
  
  if (nchar(fn_dets[2]) %in% c(3,4)){
    #kinda looks like a file
    if (file.exists(alkTable)){
      if (fn_dets[2]=="csv")custAlkTable = utils::read.csv(fn)
      if (fn_dets[2]=="xlsx")custAlkTable = openxlsx::read.xlsx(fn)
      if (fn_dets[2]=="txt")custAlkTable = utils::read.delim(fn)
    }else{
      #maybe a bad path?
      print("check Path")
    }
    
  }else{
    #not a valid file - maybe an oracle object?
    sql <- paste0("select * from ",alkTable)
    custAlkTable<-oracle_cxn$thecmd(oracle_cxn$channel, sql)
    if (class(custAlkTable)=='character')return(NA)
    if (nrow(custAlkTable)<1)return(NA)
  }
  
  return(custAlkTable)
}

R/stranal.R
#' @title stranal
#' @description stranal is a program capable of Stratified Analyses of both DFO 
#' and NMFS fisheries survey data. This package attempts to emulate the results
#' of the older, APL version which has been in use for many years.
#' 
#' Using stranal requires: 
#' \enumerate{
#' \item an Oracle account on the DFO PTRAN database;
#' \item access to PTRAN (i.e. either within the network, or via a VPN) 
#' \item appropriate select permissions on the 'groundfish' and/or 'usnefsc' 
#' schemas
#' }
#' New Oracle accounts are requested by filling out a computer account request 
#' form. These forms may be obtained from IM&TS. Select access to the DFO and 
#' NMFS data is provided by the respective datasets manager (checking with 
#' \email{Mike.McMahon@@dfo-mpo.gc.ca} is a good start).
#' 
#' Users can supply as much or as little information as they want when calling 
#' stranal(). If required information is not supplied, a select box will show 
#' the available options to the user. Please see below for examples of function 
#' calls providing both the minimum and the maximum number of parameters.
#' 
#' Fun Facts About Stranal
#' \enumerate{
#' \item Berried Females -- APL Stranal was inconsistent with species where 
#' females could be coded as 'berried' (i.e. 3).  When unsexed analyses were 
#' done, the females were included, but when sexed analyses were done, these 
#' females were not included in the results.  Such inconsistency seems 
#' suboptimal, and is retained for now such that APL and R versions of the 
#' application can be compared. 
#' \item Ages By Sex -- When APL Stranal did analyses by sex, it combined the 
#' sexes for the 'Age Mean', 'Age Total', (and standard errors of each).  This 
#' version adds the parameter \code{ageBySex} so that by setting it to FALSE, 
#' you can get the combined results, but you can also set it to TRUE, and get 
#' the values for each sex individually.
#' \item I liked the old way -- Some aspects of the 'classic' (i.e. APL) 
#' Stranal seemed inefficient.   For example, the Strata information was broken 
#' up over multiple worksheets, as was the weight information. By default, a new 
#' spreadsheet has been designed, but since there will always be those who 
#' prefer the old way, setting \code{output='classic'} will generate a 
#' spreadsheet almost indistinguishable from the APL version.
#' }
#' 
#' @param usepkg  The default value is \code{'rodbc'}, but \code{'roracle'} 
#' works as well. This describes the R package you use to connect to Oracle.  
#' @param agency  The default value is \code{'DFO'}, the other option is 
#' \code{'NMFS'}.  Setting to \code{NULL} will result in a pick list.
#' @param type This is the 'experiment type', and the default value is \code{1}.  
#' For \code{agency='DFO'}, \code{5} is also acceptable. For agency = NMFS, 136 
#' is probable, but you can type any integer, and the typed number will be 
#' embedded in\code{USNEFSC.USS_STATION.SHG <= type}. Setting to \code{NULL} 
#' will result in a pick list.
#' @param year The default value is \code{NULL}, which will result in a pick list.
#' An example of a valid, non-empty value is \code{2017}.
#' @param season The default value is \code{''}, which will result in a pick 
#' list. Valid values are \code{'SUMMER', 'SPRING', 'WINTER', and 'FALL'}, 
#' subject to the availability of data in that season for the selected agency, 
#' year, type, etc... 
#' @param missions The default value is \code{NULL}, which will result in a pick 
#' list. An example valid value is \code{c('NED2016016')}.  If your choice is 
#' not valid (given your other selections), you will be presented with a pick
#' list of valid options. 
#' @param strataTable  The default value is \code{'GROUNDFISH.GSSTRATUM'}.  
#' Depending on your data, other valid values may include: 
#' \code{'USNEFSC.DFO5ZJM','USNEFSC.DFO5ZGHNO','USNEFSC.NMFS5ZJM',
#' 'USNEFSC.NMFS5ZGHNO','USNEFSC.NMFS5ZJMC','USNEFSC.NMFS5ZJMU',
#' 'USNEFSC.NMFS5ZU'}.  Setting this to \code{NULL} will result in a pick list. 
#' @param wingspread  This is the width in feet of a standard tow.  The default 
#' value is \code{41}.  Setting this to \code{NULL} will result in a pick list. 
#' @param towDist  This is the length (NM) of a standard tow.  The default value 
#' is \code{1.75}.  Setting this to \code{NULL} will result in a pick list. 
#' @param strata  These are the strata for which you want results.  The default 
#' value is \code{c(440:495)} (DFO Summer Survey Strata).  Setting this to 
#' \code{NULL} will result in a pick list. 
#' @param spp  This is the species code for the species you want to analyze.  
#' The default value is \code{''}, which will result in a pick list. An example 
#' of a valid, non-empty value is \code{2526}. 
#' @param bySex If the selected species was measured by sex, this parameter
#' allows you to perform calculations by sex.  The default value is \code{''}, 
#' which will result in a pick list. An example of a valid, non-empty value is 
#' \code{TRUE}. 
#' @param ageBySex The APL version of stranal ignored sex differences in some 
#' age results (e.g. 'age by set', 'age mean', 'age total', ' etc.) despite
#' analyses being done by sex.  The default value of this parameter is 
#' \code{FALSE} so that the original results are emulated.  However, setting 
#' this to \code{TRUE} (when \code{bySex=TRUE}), will show age results broken 
#' down by sex.  Setting this to \code{''} will result in a pick list.
#' @param output  The default value is \code{'new'}.  This determines the format 
#' of the output Excel file. Setting this to \code{'classic'} will emulate the
#' original APL STRANAL results, including overriding your parameter for 
#' \code{ageBySex} and forcing it to FALSE.  If no excel output is desired, set
#' this parameter to an empty string \code{''}
#' @family Gale-force
#' @author  Mike McMahon, \email{Mike.McMahon@@dfo-mpo.gc.ca}
#' @importFrom RODBC odbcConnect
#' @importFrom RODBC sqlQuery
#' @importFrom stats aggregate
#' @importFrom Mar.utils make_oracle_cxn
#' @importFrom Mar.utils SQL_in
#' @importFrom Mar.utils st_err
#' @importFrom openxlsx createWorkbook
#' @importFrom openxlsx addWorksheet
#' @importFrom openxlsx writeDataTable
#' @importFrom openxlsx saveWorkbook
#' @export
#' @note Reports of errors and/or recommendations for improvements should be 
#' addressed to:
#' Mike McMahon (Mike.McMahon@dfo-mpo.gc.ca)
#' Fisheries and Oceans Canada


stranal<-function(usepkg = 'rodbc', 
                  agency = 'DFO',
                  type = 1,
                  year = NULL,
                  season = NULL,
                  missions = NULL,
                  strataTable = 'GROUNDFISH.GSSTRATUM',
                  wingspread = 41,
                  towDist = 1.75,
                  strata = c(440:495),
                  spp = NULL,
                  bySex = NULL,
                  ageBySex = FALSE,
                  output = "new"
                  ){
  if (is.null(output))output<-NA
  assign("oracle_cxn", Mar.utils::make_oracle_cxn(usepkg), envir = .GlobalEnv )
  
  agency = getUserInput("agency",agency=agency)
  type = getUserInput("type", agency=agency, type=type)
  #NED2016016
  missionsAndStrata = getUserInput("missionsAndStrata", agency=agency,type=type, 
                                   year=year, season=season, missions=missions)
    dfMissions = missionsAndStrata[[1]]
    dfMissionsStrata = missionsAndStrata[[2]]
    rm(missionsAndStrata)

  strataTable = getUserInput("strataTable", strataTable=strataTable, dfMissionsStrata=dfMissionsStrata)

  wingspread = getUserInput("wingspread", agency=agency, wingspread=wingspread)
  towDist = getUserInput("towDist", towDist=towDist)

  dfStrata = getUserInput("strata",agency=agency, strataTable=strataTable, strata = strata, 
                        dfMissionsStrata=dfMissionsStrata, towDist=towDist, 
                        wingspread=wingspread)

  spp = getUserInput("spp", agency = agency, spp=spp, bySex = bySex, 
                     ageBySex = ageBySex)
  bySex = spp[[1]]
  dfSpp = spp[[2]]
  ageBySex = spp[[3]]
    rm(spp)

  dfRawCatch <- extractData('catch', agency=agency, dfSpp=dfSpp, missions=dfMissions, strata = dfStrata$STRAT)
  dfRawInf <- extractData('inf', agency=agency, missions=dfMissions, strata = dfStrata$STRAT, type=type)
  dfRawDet <- extractData('det', agency=agency, missions=dfMissions, strata = dfStrata$STRAT, dfSpp = dfSpp, bySex = bySex, type=type)
  

  dfNWSets <- calcNumsWeights('sets',dfRawCatch=dfRawCatch,dfRawInf=dfRawInf, towDist=towDist)
  dfNWAgg <- calcNumsWeights('setsAgg', dfNWSets=dfNWSets, dfStrata=dfStrata)
  

  dfStrata <- merge(dfStrata, calcNumsWeights('strataProp', dfNWSets=dfNWSets, 
                                             dfStrata=dfStrata, dfNWAgg=dfNWAgg), all.x=T)
    allStrat = as.data.frame(dfStrata[,"STRAT"])
    colnames(allStrat)<-"STRAT"
  
  lengthsData <-calcAgeLen('lengths', agency = agency, dfNWSets=dfNWSets, dfRawDet=dfRawDet, 
                dfRawInf=dfRawInf, dfStrata=dfStrata, dfSpp=dfSpp, 
                towDist=towDist, bySex = bySex)
    agelen<-lengthsData$agelen
    lengthsTotals<-lengthsData$length_total
    lset = lengthsData$lset
  
  ageLengthKey <-calcAgeLen('ageKey', agelen=agelen, dfSpp=dfSpp, lengthsTotals 
                             = lengthsTotals, lset = lset, dfStrata=dfStrata, 
                            bySex = bySex, output = output, ageBySex = ageBySex)
  lengthsData$agelen<-NULL
  lengthsData$lset<-NULL
  metadata=list(
                "Mar.stranal" = utils::packageDescription('Mar.stranal')$Version,
                "Date" = as.character(Sys.time()),
                "Data Source" = agency,
                "Missions" = paste("'", paste(dfMissions, collapse="','"),"'", sep=""),
                "Strata" = paste("'", paste(allStrat[,1], collapse="','"),"'", sep=""),
                "Species" = paste0(dfSpp$CNAME, " (", dfSpp$SPEC ,")"),
                "By Sex" = bySex,
                "Distance" = towDist,
                "Spread" = wingspread,
                "Stratum Area Table" =  strataTable,
                "Experiment Type" = type,
                "ALK Modifications" = 'Not implemented yet')
  
  nwData <- merge(allStrat, dfNWAgg, all.x = TRUE)
    nwData[is.na(nwData)]<-0

    nwData = rbind(nwData,c("STRAT"="TOTAL",colSums(nwData[,!(colnames(nwData) =="STRAT")])))
    
  dfNWSets = merge(cbind("STRAT"= dfStrata$STRAT),dfNWSets,all.x=T)
    dfNWSets[is.na(dfNWSets)]<-0
    
  length_by_strat_mean = merge(cbind("STRAT"= dfStrata$STRAT),lengthsData$length_by_strat_mean,all.x=T)
    length_by_strat_mean[is.na(length_by_strat_mean)]<-0
    lengthsData$length_by_strat_mean<-length_by_strat_mean
    
  length_by_strat_se = merge(cbind("STRAT"= dfStrata$STRAT),lengthsData$length_by_strat_se,all.x=T)
    length_by_strat_se[is.na(length_by_strat_se)]<-0
    lengthsData$length_by_strat_se<-length_by_strat_se
    
  length_total = merge(cbind("STRAT"= dfStrata$STRAT),lengthsData$length_total,all.x=T)
    length_total[is.na(length_total)]<-0
    #add Column Totals
    length_total = rbind(length_total,c("STRAT"="TOTAL",colSums(length_total[,!(colnames(length_total) =="STRAT")])))
    lengthsData$length_total<-length_total
  length_total_se = merge(cbind("STRAT"= dfStrata$STRAT),lengthsData$length_total_se,all.x=T)
    length_total_se[is.na(length_total_se)]<-0
    lengthsData$length_total_se<-length_total_se
    
  weight_by_set = merge(dfRawInf[,c("STRAT","MISSION","SETNO")], dfNWSets[,c("STRAT","MISSION","SETNO", "TOTWGT")], all.x=TRUE)

  if (class(ageLengthKey) == "list"){
  age_total = ageLengthKey$age_total
    age_total = rbind(age_total,c("STRAT"="TOTAL",colSums(age_total[,!(colnames(age_total) =="STRAT")])))
    ageLengthKey$age_total<-age_total
  }  
  #Add the totals to Strata for approp columns
  dfStrataDataCols =c("SQNM","TUNITS","AREAPROP","AREAPROPSTERR","AREATOT","AREATOTSTERR")
  dfStrataDataTots = colSums(dfStrata[,dfStrataDataCols],na.rm = TRUE)
  names(dfStrataDataTots) <-dfStrataDataCols
  dfStrataOthCols = rep(NA, length(dfStrata[,!(colnames(dfStrata) %in% dfStrataDataCols)]))
  names(dfStrataOthCols) <- colnames(dfStrata[,!(colnames(dfStrata) %in% dfStrataDataCols)])
  totals= data.frame(t(c(dfStrataDataTots,dfStrataOthCols)))
  dfStrata = rbind(dfStrata,totals)  

  
  res=list(metadata = metadata,
           strataInfo = dfStrata,
           nwInfo = nwData,
           lengthInfo = lengthsData,
           ageInfo = ageLengthKey
  )

  if (!is.na(output)){
    wbName = "Mar_stranal.xlsx"
    md = data.frame(unlist(metadata))
    colnames(md)<-"Value"
    wb<-createWorkbook(creator = paste0("Mar.stranal v.",metadata$Mar.stranal))
    sheet1 <- addWorksheet(wb, sheetName = "QUERY")
    writeDataTable(wb, x=data.frame(md), rowNames = TRUE, sheet = sheet1, withFilter = FALSE)
    
  if (output=="classic"){
      sheet2 <- addWorksheet(wb, sheetName = "Strata Area")
        writeDataTable(wb, x=dfStrata[,c("STRAT","TUNITS","SQNM")], rowNames = FALSE, sheet = sheet2)
      sheet3 <- addWorksheet(wb, sheetName = "Prop Area")
        writeDataTable(wb, x=dfStrata[,c("STRAT","AREAPROP")], rowNames = FALSE, sheet = sheet3)
      sheet4 <- addWorksheet(wb, sheetName = "Prop Area Standard Error")
        writeDataTable(wb, x=dfStrata[,c("STRAT","AREAPROPSTERR")], rowNames = FALSE, sheet = sheet4)
      sheet5 <- addWorksheet(wb, sheetName = "Total Area")
        writeDataTable(wb, x=dfStrata[,c("STRAT","AREATOT")], rowNames = FALSE, sheet = sheet5)
      sheet6 <- addWorksheet(wb, sheetName = "Total Area Standard Error")
        writeDataTable(wb, x=dfStrata[,c("STRAT","AREATOTSTERR")], sheet = sheet6)
        if (class(ageLengthKey) == "list"){
          sheet7 <- addWorksheet(wb, sheetName = "Age Length Key")
            writeDataTable(wb, x=data.frame(ageLengthKey$alk), rowNames = TRUE, sheet = sheet7)
          sheet8 <- addWorksheet(wb, sheetName = "Age Table")
            writeDataTable(wb, x=ageLengthKey$age_table, rowNames = TRUE, sheet = sheet8)
          sheet9 <- addWorksheet(wb, sheetName = "Age Length Weight")
            writeDataTable(wb, x=ageLengthKey$alw, rowNames = TRUE, sheet = sheet9)
          sheet15 <- addWorksheet(wb, sheetName = "Age By Set")
            writeDataTable(wb, x=ageLengthKey$age_by_set, rowNames = FALSE, sheet = sheet15)
          sheet16 <- addWorksheet(wb, sheetName = "Age Mean")
            writeDataTable(wb, x=ageLengthKey$age_mean, rowNames = FALSE, sheet = sheet16)
          sheet17 <- addWorksheet(wb, sheetName = "Age Mean Standard Error")
            writeDataTable(wb, x=ageLengthKey$age_mean_se, rowNames = FALSE, sheet = sheet17)
          sheet18 <- addWorksheet(wb, sheetName = "Age Total")
            writeDataTable(wb, x=age_total, rowNames = FALSE, sheet = sheet18)
          sheet19 <- addWorksheet(wb, sheetName = "Age Total Standard Error")
            writeDataTable(wb, x=ageLengthKey$age_total_se, rowNames = FALSE, sheet = sheet19)
        } 
      sheet10 <- addWorksheet(wb, sheetName = "Length By Set")
        writeDataTable(wb, x=lengthsData$length_by_set, rowNames = FALSE, sheet = sheet10)
      sheet11 <- addWorksheet(wb, sheetName = "Length Mean")
        writeDataTable(wb, x=length_by_strat_mean, rowNames = FALSE, sheet = sheet11)
      sheet12 <- addWorksheet(wb, sheetName = "Length Mean Standard Error")
        writeDataTable(wb, x=length_by_strat_se, rowNames = FALSE, sheet = sheet12)
      sheet13 <- addWorksheet(wb, sheetName = "Length Total")
        writeDataTable(wb, x=length_total, rowNames = FALSE, sheet = sheet13)
      sheet14 <- addWorksheet(wb, sheetName = "Length Total Standard Error")
        writeDataTable(wb, x=length_total_se, rowNames = FALSE, sheet = sheet14)
      sheet20 <- addWorksheet(wb, sheetName = "Weight by Set")
        writeDataTable(wb, x=weight_by_set, rowNames = FALSE, sheet = sheet20)
      sheet21 <- addWorksheet(wb, sheetName = "Weight Mean")
        writeDataTable(wb, x=nwData[,c("STRAT","MEAN_WGT")], rowNames = FALSE, sheet = sheet21)
      sheet22 <- addWorksheet(wb, sheetName = "Weight Mean Standard Error")
        writeDataTable(wb, x=nwData[,c("STRAT","ST_ERR_WGT")], rowNames = FALSE, sheet = sheet22)
      sheet23 <- addWorksheet(wb, sheetName = "Weight Total")
        writeDataTable(wb, x=nwData[,c("STRAT","BIOMASS")], rowNames = FALSE, sheet = sheet23)
      sheet24 <- addWorksheet(wb, sheetName = "Weight Total Standard Error")
        writeDataTable(wb, x=nwData[,c("STRAT","ST_ERR_BIOMASS")], rowNames = FALSE, sheet = sheet24)
  }else if (nchar(output)>0){
    sheet2 <- addWorksheet(wb, sheetName = "Strata Info")
      writeDataTable(wb, x=dfStrata, rowNames = FALSE, sheet = sheet2)
    sheet3 <- addWorksheet(wb, sheetName = "Numbers and Weights by Strata")
      writeDataTable(wb, x=nwData, rowNames = FALSE, sheet = sheet3)
    sheet3a <- addWorksheet(wb, sheetName = "Numbers and Weights by Set")
      writeDataTable(wb, x=dfNWSets, rowNames = FALSE, sheet = sheet3a)
    if (class(ageLengthKey) == "list"){
      sheet4 <- addWorksheet(wb, sheetName = "Age Length Key")
        writeDataTable(wb, x=data.frame(ageLengthKey$alk), rowNames = TRUE, sheet = sheet4)
      sheet5 <- addWorksheet(wb, sheetName = "Age Table")
        writeDataTable(wb, x=ageLengthKey$age_table, rowNames = TRUE, sheet = sheet5)
      sheet6 <- addWorksheet(wb, sheetName = "Age Length Weight")
        writeDataTable(wb, x=ageLengthKey$alw, rowNames = TRUE, sheet = sheet6)
      sheet12 <- addWorksheet(wb, sheetName = "Age By Set")
        writeDataTable(wb, x=ageLengthKey$age_by_set, rowNames = FALSE, sheet = sheet12)
      sheet13 <- addWorksheet(wb, sheetName = "Age Mean")
        writeDataTable(wb, x=ageLengthKey$age_mean, rowNames = FALSE, sheet = sheet13)
      sheet14 <- addWorksheet(wb, sheetName = "Age Mean Standard Error")
        writeDataTable(wb, x=ageLengthKey$age_mean_se, rowNames = FALSE, sheet = sheet14)
      sheet15 <- addWorksheet(wb, sheetName = "Age Total")
        writeDataTable(wb, x=age_total, rowNames = FALSE, sheet = sheet15)
      sheet16 <- addWorksheet(wb, sheetName = "Age Total Standard Error")
        writeDataTable(wb, x=ageLengthKey$age_total_se, rowNames = FALSE, sheet = sheet16)
    }
    sheet7 <- addWorksheet(wb, sheetName = "Length By Set")
      writeDataTable(wb, x=lengthsData$length_by_set, rowNames = FALSE, sheet = sheet7)
    sheet8 <- addWorksheet(wb, sheetName = "Length Mean")
      writeDataTable(wb, x=length_by_strat_mean, rowNames = FALSE, sheet = sheet8)
    sheet9 <- addWorksheet(wb, sheetName = "Length Mean Standard Error")
      writeDataTable(wb, x=length_by_strat_se, rowNames = FALSE, sheet = sheet9)
    sheet10 <- addWorksheet(wb, sheetName = "Length Total")
      writeDataTable(wb, x=length_total, rowNames = FALSE, sheet = sheet10)
    sheet11 <- addWorksheet(wb, sheetName = "Length Total Standard Error")
    writeDataTable(wb, x=length_total_se, rowNames = FALSE, sheet = sheet11)
  } 
    saveWorkbook(wb, file = wbName, overwrite=TRUE)
    cat(paste0("\n\nWrote your excel file to ",file.path(getwd(),wbName),""))
  }
 
  return(res)
  }
